### PR TITLE
Tests for appsody build

### PR DESCRIPTION
**Pull request by upLukeWeston**
Resolves potions of #699.

- Test with an invalid `tag` associated with `push-url`
- Test with an invalid `push-url`
- Test for checking that `app-deploy-yaml` already exists before writing a new one

### CODE REVIEW

File: functest/build_test.go [hunk 1]
```diff
@@ -413,36 +413,78 @@ func verifyImageAndConfigLabelsMatch(t *testing.T, deploymentManifest cmd.Deploy
...
+				// If an error is not returned, the test should fail
+			} else {
+				t.Error("Build with missing tag did not fail as expected")
+			}
+		})
+	}
 }
 
-func TestBuildMissingTagFail(t *testing.T) {
+func TestBuildConfigAlreadyExists(t *testing.T) {
```
Comment (EnriqueL8) [inline functest/build_test.go]: Would just renamed this to `TestBuildDeploymentConfigAlreadyExists`

File: functest/build_test.go [hunk 2]
```diff
@@ -413,36 +413,78 @@ func verifyImageAndConfigLabelsMatch(t *testing.T, deploymentManifest cmd.Deploy
...
+func TestInvalidBuild(t *testing.T) {
+
+	var knativeFlagTests = []struct {
+		testName    string
+		args        []string
+		expectedLog string
+	}{
+		{"Missing Tag ", []string{"--push"}, "Cannot specify --push or --push-url without a --tag"},
+		{"Invalid Tag with Push URL", []string{"--push-url", "i.am.not.a.real.url", "--tag", "£"}, "invalid argument \"i.am.not.a.real.url/£\" for \"-t, --tag"},
+		{"Invalid Push URL", []string{"--push-url", "i.am.not.a.real.url", "--tag", "notgonna/work"}, "Could not push the docker image"},
```
Comment (EnriqueL8) [inline functest/build_test.go]: Maybe the message here should be invalid push url? could be difficult to verify that the url is valid.

Comment (upLukeWeston) [reply functest/build_test.go]: I'm not sure which message you're referring to? The message is already 'invalid push url'?

Comment (EnriqueL8) [reply functest/build_test.go]: I mean the expect log being `Could not push the docker image`

File: cmd/build.go [hunk 3]
```diff
@@ -155,19 +155,21 @@ func build(config *buildCommandConfig) error {
...
+		// check to see if the pushURL provided is in the expected format
+		// Valid examples would be:
+		// https://www.appsody.co.uk
+		// http://www.appsody.co.uk
+		// www.appsody.co.uk
+		// appsody.co.uk
+		matched, err := regexp.MatchString("^(http:\\/\\/www\\.|https:\\/\\/www\\.|http:\\/\\/|https:\\/\\/)?[a-z0-9]+([\\-\\.]{1}[a-z0-9]+)*\\.[a-z]{2,5}(:[0-9]{1,5})?(\\/.*)?$", config.pushURL)
+		if err != nil {
+			return err
+		}
```
Comment (EnriqueL8) [inline cmd/build.go]: We would need to put this function in utils, and write specific tests for it.I tested it against `default-route-openshift-image-registry.apps-crc.testing` which should match but it didn't. I think maybe we can put this change in a different PR and merge the build tests for now?

Comment (upLukeWeston) [reply cmd/build.go]: I was thinking about this yeah, but wanted to keep it separate from the `build_test` stuff.  I'll extrapolate it into a function and write some test cases for it.

**Comment by skoh7645** [comment]
Following @EnriqueL8's review, this LGTM
